### PR TITLE
Try to CI setup for Windows 

### PR DESCRIPTION
**Pull request by jeffin143**
This may be a big failure 
I have no idea about windows

### CODE REVIEW

File: .ci/windows-steps.yaml [hunk 1]
```diff
@@ -0,0 +1,169 @@
...
+    cp $(Agent.ToolsDirectory)\OpenBLAS.0.2.14.1\lib\native\lib\x64\*.* build\
+    cp $(Agent.ToolsDirectory)\OpenBLAS.0.2.14.1\lib\native\bin\x64\*.* build\
+    cd build
+    cmake $(CMakeGenerator) `
+    -DBLAS_LIBRARIES:FILEPATH=$(Agent.ToolsDirectory)\OpenBLAS.0.2.14.1\lib\native\lib\x64\libopenblas.dll.a `
+    -DLAPACK_LIBRARIES:FILEPATH=$(Agent.ToolsDirectory)\OpenBLAS.0.2.14.1\lib\native\lib\x64\libopenblas.dll.a `
+    -DARMADILLO_INCLUDE_DIR="..\armadillo-8.400.0\include" `
+    -DARMADILLO_LIBRARY="..\armadillo-8.400.0\Release\armadillo.lib" `
+    -DProtobuf_INCLUDE_DIR="..\protobuf\src\" `
+    -DPROTOBUF_LIBRARY="..\protobuf\cmake\solution\Release\libprotobuf.lib" `
```
Comment (zoq) [inline .ci/windows-steps.yaml]: I guess you already checked that `libprotobuf.lib` is in `..\protobuf\cmake\solution\Release\libprotobuf.lib` before the build?

Might be  good to copy `libprotobuf.*` into `build\` as well make sure it's there and use `$(Agent.ToolsDirectory)\boost_libs\libprotobuf.lib` instead.

Debugging a windows build is often not as easy.

File: .ci/windows-steps.yaml [hunk 2]
```diff
@@ -0,0 +1,170 @@
...
+
+# Build mlpack
+- task: MSBuild@1
+  inputs:
+    solution: 'mlpack/build/*.sln'
+    msbuildLocationMethod: 'location'
+    msbuildVersion: $(MSBuildVersion)
+    configuration: 'Release'
+    msbuildArchitecture: 'x64'
+    msbuildArguments: /m /p:BuildInParallel=true
```
Comment (zoq) [inline .ci/windows-steps.yaml:96]: Perhaps adding `/MD` to each Build section solved the current build issue, not sure.

File: tests/main.cpp [hunk 3]
```diff
@@ -5,47 +5,46 @@
 
 #include <iostream>
 #include <mlboard/core.hpp>
-#include <unistd.h>
 #include <dirent.h>
```
Comment (jeffin143) [inline tests/main.cpp:8]: These two header files are not there in windows (Or to be specific visual studio ) so the code won't work

Note : it is there in mingw

File: tests/catch.hpp [hunk 4]
```diff
@@ -8071,7 +8071,7 @@ namespace Generators {
 
     template<>
     auto all<int>() -> Generator<int> {
-        return range( std::numeric_limits<int>::min(), std::numeric_limits<int>::max() );
+        return range( (std::numeric_limits<int>::min)(), (std::numeric_limits<int>::max)() );
```
Comment (birm) [inline tests/catch.hpp]: Hmm. It may be worthwhile to report the issue you got which lead to this upstream.

File: tests/filewriter_test.cpp [hunk 5]
```diff
@@ -4,10 +4,14 @@
  */
 #include "catch.hpp"
 #include <mlboard/mlboard.hpp>
-#include <unistd.h>
 #include <sstream>
 #include <cstdio>
 #include <sys/stat.h>
+
+// for windows mkdir
```
Comment (birm) [inline tests/filewriter_test.cpp]:
```suggestion
// For windows mkdir.
```

File: tests/main.cpp [hunk 6]
```diff
@@ -5,13 +5,19 @@
...
 
 //#define CATCH_CONFIG_MAIN  // catch.hpp will define main().
 #define CATCH_CONFIG_RUNNER  // We will define main().
 #include "catch.hpp"
 
+// for windows rmdir
```
Comment (birm) [inline tests/main.cpp]:
```suggestion
// For windows rmdir.
```

**Review by birm — APPROVED** [review]
Seems strange to me how big the variance in CI times is, but it seems to be working just fine. Great work!

File: tests/catch.hpp (hunk 4 above)
Comment (jeffin143) [reply tests/catch.hpp]: The fixed it upstream, I took the version from ensmallen and hence that was old and having issue, but it is resolved in the version which I changed it

File: tests/filewriter_test.cpp [hunk 7]
```diff
@@ -28,9 +32,15 @@ TEST_CASE("Writing two files at a time", "[FileWriter]")
...
+        _mkdir("_temp1_");
+        _mkdir("_temp2_");
+    #else 
+        mkdir("_temp1_",0777);
+        mkdir("_temp2_",0777);
+    #endif
     // Create temp dirs.
-    mkdir("_temp1_",0777);
-    mkdir("_temp2_",0777);
+    // failing
```
Comment (toshal-a) [inline tests/filewriter_test.cpp]:
```suggestion

```

File: tests/filewriter_test.cpp [hunk 8]
```diff
@@ -28,9 +32,15 @@ TEST_CASE("Writing two files at a time", "[FileWriter]")
...
 TEST_CASE("Writing two files at a time in different paths", "[FileWriter]")
 {
+    #if defined(_WIN32)
+        _mkdir("_temp1_");
+        _mkdir("_temp2_");
+    #else 
+        mkdir("_temp1_",0777);
+        mkdir("_temp2_",0777);
+    #endif
     // Create temp dirs.
```
Comment (toshal-a) [inline tests/filewriter_test.cpp]:
```suggestion
    // Create temp dirs.
    #if defined(_WIN32)
        _mkdir("_temp1_");
        _mkdir("_temp2_");
    #else 
        mkdir("_temp1_",0777);
        mkdir("_temp2_",0777);
    #endif
```

File: tests/filewriter_test.cpp [hunk 9]
```diff
@@ -53,7 +63,13 @@ TEST_CASE("Writing two files at a time in different paths", "[FileWriter]")
 TEST_CASE("Writing a summary to file", "[FileWriter]")
 {
     // Create temp dirs.
-    mkdir("_temp1_",0777);
+
```
Comment (toshal-a) [inline tests/filewriter_test.cpp]:
```suggestion
```

File: tests/main.cpp [hunk 10]
```diff
@@ -5,13 +5,19 @@
...
 #define CATCH_CONFIG_RUNNER  // We will define main().
 #include "catch.hpp"
 
+// For windows rmdir.
+#ifdef _WIN32
+  #include <io.h>
+  #include <direct.h>
+#endif
+
+
```
Comment (toshal-a) [inline tests/main.cpp]:
```suggestion
```

**Review by toshal-a — APPROVED** [review]
Hi, I have left some comments.These would be probably final comments.

**Comment by birm** [comment]
@walragatver you said "probably" so I want to give you a chance to comment before merging.

**Pull request merged**